task/tkn-bundle-oci-ta/0.2/tkn-bundle-oci-ta.yaml

```diff
@@ -191,9 +195,92 @@ spec:
+
+          # Check if task bundle has an attached migration file.
+          local filename
+          local found=
+          local artifact_refs
+
+          # List attached artifacts, that have specific artifact type and annotation.
+          # Then, find out the migration artifact.
+          #
+          # Minimum version oras 1.2.0 is required for option --format
+          artifact_refs=$(
+            oras discover "$task_bundle" --artifact-type "$ARTIFACT_TYPE_TEXT_XSHELLSCRIPT" --format json |
+              jq -r "
+              .manifests[]
+              | select(.annotations.\"${ANNOTATION_IS_MIGRATION}\" == \"true\")
+              | .reference"
+          )
+          while read -r artifact_ref; do
+            if [ -z "$artifact_ref" ]; then
+              continue
+            fi
+            filename=$(
+              retry oras pull --format json "$artifact_ref" | jq -r "
+                .files[]
+                | select(.annotations.\"org.opencontainers.image.title\" == \"${task_version}.sh\")
+                | .annotations.\"org.opencontainers.image.title\"
+                "
+            )
+
+            if [ -n "$filename" ]; then
+              if diff "$filename" "$migration_file" >/dev/null; then
+                found=true
+                break
+              else
+                echo "error: task bundle $task_bundle has migration artifact $artifact_ref, but the migration content is different: $filename" 1>&2
+                exit 1
+              fi
+            fi
+          done <<<"$artifact_refs"
+
+          # the same migration file exist
+          if [ "$found" == "true" ]; then
+            echo 1
+          fi
+          # migratioan file doesn't exist
+          echo 0
+        }
+
+        attach_migration_file() {
+          local -r task_bundle=$1
+          local -r task_version=$2
+          local -r migration_file=$3
+          local -r auth_json=$4
+
+          cd "${migration_file%/*}"
+          retry oras attach \
+            --registry-config "${auth_json}" \
+            --artifact-type "$ARTIFACT_TYPE_TEXT_XSHELLSCRIPT" \
+            --annotation "${ANNOTATION_IS_MIGRATION}=true" \
+            "$task_bundle" "${migration_file##*/}"
+
+          local status=$?
+          if [[ $status -ne 0 ]]; then
+            echo "failed to attach migration script file to $task_bundle"
+            echo 1
+          fi
+
+          echo
+          echo "Attached migration file $migration_file to $task_bundle"
+
+          echo 0
+        }
+
         # task_dir is where all the tasks definitions reside
         # all tasks definitions are in the same task_dir, so extracting it from the first element is enough
-        task_dir=$(echo "${FILES[0]}" | awk -F'/' '{NF--; print $0}' OFS='/')
+        read -r task_dir task_name task_version <<<"$(
+          dirname "${FILES[0]}" |
+            awk -F'/' '{ print $0, $6, $7 }'
+        )"
+        echo "info: build and push task $task_dir" 1>&2
```

Comment:
What is the reason behind redirecting this info message to stderr?

Reply:
I copy it from build-and-push.sh, and feel it is just to collect error info/debug? Let me remove it.

Reply:
Fixed.

Reply:
Fixed.